File: server/python/tests/conformance/conformance_adapter.py
"""Bridge the corpus to the Python loader."""
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Optional

from metaobjects import LoadResult, MetaDataLoader
from metaobjects.core_types import core_provider
from metaobjects.documentation import doc_provider
from metaobjects.errors import ErrorCode, ParseError
from metaobjects.meta.meta_data import MetaData
from metaobjects.meta.template.meta_template import MetaTemplate
from metaobjects.provider import Provider
from metaobjects.registry import AttrSchema, ChildRule, TypeDefinition
from metaobjects.serializer_json import canonical_serialize
from metaobjects.shared.base_types import TYPE_ATTR
from metaobjects.source.error_source import (
    CodeSource,
    JsonSource,
    MergedSource,
    ResolvedSource,
    YamlSource,
)

# ---------------------------------------------------------------------------
# Test-only providers exercised by the provider-extension-* fixtures.
# Mirrors the TS adapter (metadata/test/conformance/adapter.ts) and the C#
# adapter (ConformanceAdapter.cs) shape for cross-port parity.
# ---------------------------------------------------------------------------


def _example_template_briefing_provider() -> Provider:
    """Adds a hypothetical `template.briefing` subtype. Fictional — used only
    by the provider-extension-* fixtures to exercise registry.register without
    colliding with real core subtypes. (Pre-ADR-0011 this was a "toolcall"
    subtype; toolcall is now core, so the test-only one moved to a clearly-
    fictional name.)
    """
    p = Provider(
        "example-template-briefing",
        dependencies=("metaobjects-core-types",),
    )
    p.add(
        TypeDefinition(
            type="template",
            sub_type="briefing",
            factory=lambda t, s, n: MetaTemplate(t, s, n),
            attrs=[
                AttrSchema("payloadRef", "string", required=True),
                AttrSchema("author", "string", required=True),
                AttrSchema("recipient", "string", required=True),
            ],
            child_rules=[ChildRule(TYPE_ATTR, "*")],
        )
    )
    return p


def _noop_provider(provider_id: str, *deps: str) -> Provider:
    return Provider(provider_id, dependencies=tuple(deps))


# Provider-id -> provider. The fixture corpus names providers by their stable
# id (e.g. "metaobjects-core-types"). load_fixture_* resolve those ids to the
# actual provider objects to compose a registry. Test-only providers
# (provider-extension-* fixtures) live alongside the shipped ones; the
# fixture's providers.json acts as the gate.
_PROVIDER_MAP: dict[str, Provider] = {
    core_provider.id: core_provider,                # "metaobjects-core-types"
    doc_provider.id: doc_provider,                  # "metaobjects-documentation"
    # Test-only — provider-extension-* fixtures.
    "example-template-briefing": _example_template_briefing_provider(),
    "cycle-a": _noop_provider("cycle-a", "cycle-b"),
    "cycle-b": _noop_provider("cycle-b", "cycle-a"),
    "depends-on-missing": _noop_provider("depends-on-missing", "does-not-exist"),
    "duplicate-x": _noop_provider("duplicate-x"),
    # Same `.id` as duplicate-x — surfaces ERR_PROVIDER_DUPLICATE_ID at compose time.
    "duplicate-x-clone": _noop_provider("duplicate-x"),
}


def _resolve_providers(provider_ids: Optional[Iterable[str]]) -> list[Provider]:
    """Map fixture-declared provider ids to provider objects.

    When ``provider_ids`` is None, falls back to the legacy default
    (``[core_provider, doc_provider]``) — pre-rc.3 behavior. An unknown id
    raises (parity with the TS / C# adapters).
    """
    if provider_ids is None:
        return [core_provider, doc_provider]
    resolved: list[Provider] = []
    for pid in provider_ids:
        p = _PROVIDER_MAP.get(pid)
        if p is None:
            raise ValueError(f'Unknown provider id "{pid}"')
        resolved.append(p)
    return resolved


@dataclass(frozen=True)
class ErrorEnvelopeRecord:
    """FR5a / ADR-0009 — cross-port envelope record surfaced by the adapter.

    Mirrors the TS ``ErrorEnvelopeRecord`` shape so the Python conformance
    runner can do the same per-error envelope assertion the TS runner does.
    FR5d — ``referrer`` and ``target`` are populated for ``format=resolved``
    envelopes (reference-resolution errors).
    """
    code: str
    format: str
    files: tuple[str, ...]
    json_path: Optional[str]
    referrer: Optional[str] = None
    target: Optional[str] = None


def _relativize(file_path: str, input_dir: Path) -> str:
    """Make a file path relative to the fixture's input directory."""
    p = Path(file_path)
    try:
        rel = p.absolute().relative_to(input_dir.absolute())
        return str(rel).replace("\\", "/")
    except ValueError:
        # Already relative or under a different root.
        return file_path.replace("\\", "/")


def _build_envelope_from_source(code: str, env, input_dir: Path) -> ErrorEnvelopeRecord:
    """Build an envelope record from a (code, ErrorSource) pair.

    Shared between error-side and warning-side normalization so both channels
    produce structurally identical records (mirrors the TS adapter's single
    normalization path).
    """
    def rel_files() -> tuple[str, ...]:
        return tuple(_relativize(f, input_dir) for f in env.files)

    if isinstance(env, JsonSource):
        return ErrorEnvelopeRecord(code, "json", rel_files(), env.json_path)
    if isinstance(env, YamlSource):
        return ErrorEnvelopeRecord(code, "yaml", rel_files(), env.json_path)
    if isinstance(env, MergedSource):
        return ErrorEnvelopeRecord(code, "merged", rel_files(), env.json_path)
    if isinstance(env, ResolvedSource):
        # FR5d — surface referrer + target so the cross-port runner can assert them.
        return ErrorEnvelopeRecord(
            code, "resolved", rel_files(), env.json_path,
            referrer=env.referrer, target=env.target,
        )
    if isinstance(env, CodeSource):
        return ErrorEnvelopeRecord(code, "code", (), None)
    # No envelope — synthesize a minimal root-level shape.
    return ErrorEnvelopeRecord(code, "json", (), "$")


def _build_envelope(err, input_dir: Path) -> ErrorEnvelopeRecord:
    """Convert a Python MetaError into the cross-port envelope record."""
    return _build_envelope_from_source(err.code.name, err.envelope, input_dir)


def _build_warning_envelope(warn, input_dir: Path) -> ErrorEnvelopeRecord:
    """Convert a Python LoaderWarning into the cross-port envelope record.

    FR5c-finalize — mirrors :func:`_build_envelope` so the cross-port runner
    can assert warning envelopes the same way it asserts error envelopes.
    """
    return _build_envelope_from_source(warn.code, warn.source, input_dir)


def _load_or_capture_compose_error(
    input_dir: Path,
    providers: list[Provider],
) -> tuple[Optional[LoadResult], Optional[str]]:
    """Try to compose + load. If composition throws, return (None, code_str).
    Otherwise return (result, None). Composition errors surface as
    code-only outcomes — parity with the TS and C# adapters.
    """
    try:
        result = MetaDataLoader.from_directory(input_dir, providers=providers)
        return result, None
    except ParseError as ex:
        # composeRegistry raises ParseError with ERR_PROVIDER_* codes.
        if ex.code in (
            ErrorCode.ERR_PROVIDER_DUPLICATE_ID,
            ErrorCode.ERR_PROVIDER_MISSING_DEPENDENCY,
            ErrorCode.ERR_PROVIDER_DEPENDENCY_CYCLE,
        ):
            return None, ex.code.name
        raise


def load_fixture(
    input_dir: Path,
    provider_ids: Optional[Iterable[str]] = None,
) -> tuple[list[str], list[str], str]:
    """Return (error_codes, warnings, canonical_serialization).

    ``provider_ids`` — declared by the fixture's providers.json. ``None``
    preserves pre-rc.3 behavior (``[core_provider, doc_provider]``).
    """
    providers = _resolve_providers(provider_ids)
    result, compose_code = _load_or_capture_compose_error(input_dir, providers)
    if compose_code is not None:
        return [compose_code], [], ""
    assert result is not None
    codes = [e.code.name for e in result.errors]
    canonical = canonical_serialize(result.root)
    return codes, list(result.warnings), canonical


def load_fixture_with_envelopes(
    input_dir: Path,
    provider_ids: Optional[Iterable[str]] = None,
) -> tuple[
    list[str],
    list[ErrorEnvelopeRecord],
    list[str],
    list[ErrorEnvelopeRecord],
    str,
]:
    """Return (error_codes, error_envelopes, warnings, warning_envelopes, canonical).

    FR5a / ADR-0009 — provides the per-error envelope alongside the legacy
    code-list so the conformance runner can do the envelope assertion.

    FR5c-finalize — additionally provides per-warning envelopes (same
    normalization as the error side) so the runner can assert warning
    envelope shape when ``expected-errors.json#warnings[]`` declares it.

    ``provider_ids`` — declared by the fixture's providers.json. ``None``
    preserves pre-rc.3 behavior (``[core_provider, doc_provider]``).
    """
    providers = _resolve_providers(provider_ids)
    result, compose_code = _load_or_capture_compose_error(input_dir, providers)
    if compose_code is not None:
        # Synthesize a code-only outcome — same shape the TS and C# adapters
        # surface on a composition failure (CodeSource envelope).
        compose_env = ErrorEnvelopeRecord(compose_code, "code", (), None)
        return [compose_code], [compose_env], [], [], ""
    assert result is not None
    codes = [e.code.name for e in result.errors]
    envelopes = [_build_envelope(e, input_dir) for e in result.errors]
    warning_envelopes = [
        _build_warning_envelope(w, input_dir) for w in result.get_envelope_warnings()
    ]
    canonical = canonical_serialize(result.root)
    return codes, envelopes, list(result.warnings), warning_envelopes, canonical


def load_fixture_result(
    input_dir: Path,
    provider_ids: Optional[Iterable[str]] = None,
) -> LoadResult:
    """Return the full LoadResult (including the root node) for script.json checks.

    ``provider_ids`` — declared by the fixture's providers.json. ``None``
    preserves pre-rc.3 behavior (``[core_provider, doc_provider]``).
    """
    providers = _resolve_providers(provider_ids)
    return MetaDataLoader.from_directory(input_dir, providers=providers)

File: server/csharp/MetaObjects.Conformance.Tests/ConformanceAdapter.cs
// ConformanceAdapter — C# port adapter.
//
// Ports the contract of conformance/src/adapter.ts + test/conformance/adapter.ts.
// Maps provider ids to IMetaDataTypeProvider objects, composes the registry,
// runs the loader, and exposes canonical serialization.
//
// Navigate / Invoke (capability script) are added in Slice 7.

using MetaObjects;
using MetaObjects.Core.Documentation;
using MetaObjects.Loader;
using MetaObjects.Meta;
using MetaObjects.Source;
using MetaObjects.Shared;

namespace MetaObjects.Conformance.Tests;

// ---------------------------------------------------------------------------
// Test-only providers exercised by the provider-extension-* fixtures.
//
// These are NOT shipped — they live in test code to verify the cross-port
// composition contract end-to-end. Each adapter (TS / C# / Python) carries
// the same set so fixture-declared provider ids resolve identically.
// ---------------------------------------------------------------------------

// Adds a hypothetical `template.briefing` subtype with @payloadRef + @author
// + @recipient. Fictional — used only by the provider-extension-* fixtures
// to exercise registry.Register without colliding with real core subtypes.
// (Pre-ADR-0011 this was a "toolcall" subtype; toolcall is now core, so the
// test-only one moved to a clearly-fictional name.)
internal sealed class ExampleTemplateBriefingProvider : IMetaDataTypeProvider
{
    public string Id => "example-template-briefing";
    public IReadOnlyList<string> Dependencies => new[] { "metaobjects-core-types" };

    public void RegisterTypes(TypeRegistry registry)
    {
        registry.Register(new TypeDefinition(
            typeId: new TypeId("template", "briefing"),
            description: "Hypothetical briefing template — test-only.",
            childRules: new List<ChildRule>
            {
                new ChildRule(
                    BaseTypes.TYPE_ATTR,
                    Structural.CHILD_RULE_WILDCARD,
                    Structural.CHILD_RULE_WILDCARD),
            },
            factory: (typeId, name) => new MetaTemplate(typeId, name),
            attributes: new List<AttrSchema>
            {
                new AttrSchema("payloadRef", "string", Required: true, Description: "Briefing-input payload reference."),
                new AttrSchema("author",     "string", Required: true, Description: "Author of the briefing."),
                new AttrSchema("recipient",  "string", Required: true, Description: "Intended recipient role."),
            }));
    }
}

internal sealed class NoopTestProvider(string id, params string[] deps) : IMetaDataTypeProvider
{
    public string Id { get; } = id;
    public IReadOnlyList<string> Dependencies { get; } = deps;
    public void RegisterTypes(TypeRegistry registry) { }
}

/// <summary>
/// Cross-port envelope record surfaced by <see cref="LoadOutcome.Errors"/>.
/// Mirrors the TS <c>ErrorEnvelopeRecord</c> shape so the C# runner can do
/// the same per-error envelope assertion the TS runner does. FR5d —
/// <c>Referrer</c> and <c>Target</c> are populated for <c>format=resolved</c>
/// envelopes (reference-resolution errors).
/// </summary>
public sealed record ErrorEnvelopeRecord(
    string Code,
    string Format,
    IReadOnlyList<string> Files,
    string? JsonPath,
    string? Referrer = null,
    string? Target = null);

/// <summary>
/// Result of loading a fixture's input directory.
/// </summary>
/// <param name="Tree">The loaded metadata tree.</param>
/// <param name="ErrorCodes">Error code strings collected during loading.</param>
/// <param name="Warnings">Warning strings collected during loading.</param>
/// <param name="Errors">Full envelope records (FR5a / ADR-0009). Populated alongside <see cref="ErrorCodes"/>.</param>
/// <param name="WarningEnvelopes">
/// FR5c-finalize — full warning envelopes (same shape as <see cref="Errors"/>).
/// When present AND the fixture's expected-errors.json declares envelope-shape
/// warnings, the runner asserts per-warning envelope alignment (mirrors the
/// error-side assertion). When absent, the runner falls back to comparing
/// <see cref="Warnings"/> as a flat string list against expected-warnings.json.
/// </param>
public sealed record LoadOutcome(
    MetaRoot Tree,
    IReadOnlyList<string> ErrorCodes,
    IReadOnlyList<string> Warnings,
    IReadOnlyList<ErrorEnvelopeRecord> Errors,
    IReadOnlyList<ErrorEnvelopeRecord> WarningEnvelopes);

/// <summary>
/// Adapter that bridges the conformance test infrastructure to the C# MetaObjects library.
/// </summary>
public static class ConformanceAdapter
{
    /// <summary>
    /// Provider-id → provider. The fixture corpus names providers by stable id;
    /// this maps them to provider objects. An unknown id throws (parity with the TS adapter).
    /// Test-only providers (provider-extension-* fixtures) live alongside the
    /// shipped ones — they are guarded by the fixture's providers.json so they
    /// never feed a production code path.
    /// </summary>
    private static readonly IReadOnlyDictionary<string, IMetaDataTypeProvider> Providers =
        new Dictionary<string, IMetaDataTypeProvider>(StringComparer.Ordinal)
        {
            ["metaobjects-core-types"]    = CoreTypes.CoreTypesProvider,
            ["metaobjects-documentation"] = DocumentationTypes.DocTypesProvider,
            // R6 Plan 2b — the shared db-attr fixtures (e.g. @dbColumnType pairing)
            // declare a "metaobjects-db" provider. In C# the physical-attr handling +
            // @dbColumnType pairing validation live in the always-on loader passes
            // (ValidationPasses.ValidateDbColumnType), not a separate registered
            // provider, so the id maps to a no-op composition entry whose only job is
            // to resolve the fixture's providers.json. The validation fires regardless.
            ["metaobjects-db"]            = new NoopTestProvider("metaobjects-db", "metaobjects-core-types"),
            // Test-only — provider-extension-* fixtures.
            ["example-template-briefing"] = new ExampleTemplateBriefingProvider(),
            ["cycle-a"]                   = new NoopTestProvider("cycle-a", "cycle-b"),
            ["cycle-b"]                   = new NoopTestProvider("cycle-b", "cycle-a"),
            ["depends-on-missing"]        = new NoopTestProvider("depends-on-missing", "does-not-exist"),
            ["duplicate-x"]               = new NoopTestProvider("duplicate-x"),
            // Same `.Id` as duplicate-x — surfaces ERR_PROVIDER_DUPLICATE_ID at compose time.
            ["duplicate-x-clone"]         = new NoopTestProvider("duplicate-x"),
        };

    /// <summary>
    /// Load a fixture's input directory using the specified provider ids.
    /// </summary>
    public static LoadOutcome LoadFixture(string inputDir, IReadOnlyList<string> providers)
    {
        var resolved = providers
            .Select(id => Providers.TryGetValue(id, out var p)
                ? p
                : throw new ArgumentException($"Unknown provider id \"{id}\""))
            .ToList();

        // Provider composition errors (duplicate id / missing dep / cycle) are
        // first-class fixture outcomes: surface them as a code-only LoadOutcome
        // so the cross-port runner can compare expected-errors.json directly.
        TypeRegistry registry;
        try
        {
            registry = Provider.ComposeRegistry(resolved);
        }
        catch (MetaModelException ex)
        {
            var codeStr = ex.Code.ToString();
            var compErr = new ErrorEnvelopeRecord(codeStr, "code", Array.Empty<string>(), null);
            return new LoadOutcome(
                Tree: null!,
                ErrorCodes: new[] { codeStr },
                Warnings: Array.Empty<string>(),
                Errors: new[] { compErr },
                WarningEnvelopes: Array.Empty<ErrorEnvelopeRecord>());
        }
        var result = MetaDataLoader.FromDirectory(inputDir, registry);

        // FR5a — surface the full envelope per error. Normalize files[] to
        // be relative to inputDir (the harness's portable file token).
        var envelopes = result.Errors.Select(e => BuildEnvelope(e, inputDir)).ToList();

        // FR5c-finalize — surface warning envelopes (same envelope shape as
        // errors). The loader's EnvelopeWarnings list carries envelope-shaped
        // warnings (e.g. WARN_DUPLICATE_DECLARATION) with full LoaderWarning
        // provenance. Mirror the error-side normalization so the cross-port
        // runner can assert per-warning envelope alignment.
        var warningEnvelopes = (result.EnvelopeWarnings ?? Array.Empty<LoaderWarning>())
            .Select(w => BuildWarningEnvelope(w, inputDir))
            .ToList();

        return new LoadOutcome(
            result.Root,
            result.Errors.Select(e => e.Code.ToString()).ToList(),
            result.Warnings.ToList(),
            envelopes,
            warningEnvelopes);
    }

    private static ErrorEnvelopeRecord BuildEnvelope(MetaError err, string inputDir) =>
        BuildEnvelopeFromSource(err.Code.ToString(), err.Envelope, inputDir);

    private static ErrorEnvelopeRecord BuildWarningEnvelope(LoaderWarning w, string inputDir) =>
        BuildEnvelopeFromSource(w.Code, w.Source, inputDir);

    private static ErrorEnvelopeRecord BuildEnvelopeFromSource(string code, ErrorSource? envelope, string inputDir)
    {
        IReadOnlyList<string> Rel(IEnumerable<string> files) =>
            files.Select(f => RelativizeFile(f, inputDir)).ToList();
        return envelope switch
        {
            JsonSource js     => new ErrorEnvelopeRecord(code, "json",     Rel(js.Files), js.JsonPath),
            YamlSource ys     => new ErrorEnvelopeRecord(code, "yaml",     Rel(ys.Files), ys.JsonPath),
            MergedSource ms   => new ErrorEnvelopeRecord(code, "merged",   Rel(ms.Files), ms.JsonPath),
            // FR5d — surface Referrer + Target so the cross-port runner can assert them.
            ResolvedSource rs => new ErrorEnvelopeRecord(code, "resolved", Rel(rs.Files), rs.JsonPath, rs.Referrer, rs.Target),
            CodeSource        => new ErrorEnvelopeRecord(code, "code",     new List<string>(), null),
            // Pre-FR5a fallback: no envelope. Synthesize a minimal $-rooted JSON shape.
            _                 => new ErrorEnvelopeRecord(code, "json",     new List<string>(), "$"),
        };
    }

    private static string RelativizeFile(string filePath, string inputDir)
    {
        if (filePath.StartsWith(inputDir, StringComparison.Ordinal))
        {
            var rel = System.IO.Path.GetRelativePath(inputDir, filePath);
            return rel.Replace('\\', '/');
        }
        return filePath.Replace('\\', '/');
    }

    /// <summary>
    /// Produce the canonical serialization of a loaded tree.
    /// </summary>
    public static string CanonicalSerialize(MetaRoot tree) =>
        SerializerJson.CanonicalSerialize(tree);

    /// <summary>
    /// Produce the effective (super-resolved) canonical serialization of a loaded tree.
    /// </summary>
    public static string CanonicalSerializeEffective(MetaRoot tree) =>
        SerializerJson.CanonicalSerializeEffective(tree);

    /// <summary>
    /// Navigate a path from the tree root, returning the resolved node or
    /// <see langword="null"/> if any segment does not match.
    /// </summary>
    public static MetaData? Navigate(MetaRoot tree, IReadOnlyList<string> path) =>
        Navigator.Navigate(tree, path);

    /// <summary>
    /// Invoke a capability on a node. Throws <see cref="UnknownCapabilityException"/>
    /// for an unbound capability-id.
    /// </summary>
    public static NormalizedResult Invoke(
        MetaData node,
        string capabilityId,
        IReadOnlyDictionary<string, object?> args) =>
        CapabilityBinding.Invoke(node, capabilityId, args);
}
